### PR TITLE
fix: add fixes for the updater and the match regex

### PR DESCRIPTION
**Pull request by chrisneven**
Hello @amannn,

Here I am again 😬. So I've noticed some more bugs which I create this PR for.

Firstly, `search` and `router.query` that were passed to the `url` parameter conflicted. When clearing the params for example, they'd remain in the `router.query` object and got basically out of sync with the `useQueryParam(s)` hook. I modified the `updater` function so that it passes only the `search`.

Furthermore, I've modified the match regex in order to also consider hashes in the `asPath` as there was still some funkyness going on when switching up and changing params.

fyi, I also tried it with a dynamic route and this works as expected.

### CODE REVIEW

File: packages/next-query-params/package.json [hunk 1]
```diff
@@ -22,7 +22,7 @@
     "start": "tsdx watch",
     "build": "tsdx build",
     "test": "echo \"No tests yet\"",
-    "lint": "eslint src && tsc",
+    "lint": "eslint src --fix && tsc",
```
Comment (amannn) [inline packages/next-query-params/package.json:25]: Please revert this change, I'd want CI to fail when the linter finds problems.

**Review by amannn — COMMENTED** [review]
Hey @chrisneven! Nice to hear from you again 😁.

I've just set up some Cypress testing for the example on the main branch, as I think this becomes important now to get the details right and to avoid regressions. Do you think you could rebase your branch and add tests that illustrate the behaviour and would fail before? Thanks 🙌

**Comment by chrisneven** [comment]
Hi @amannn, 

I think cypress is a really nice addition, thanks for adding it. I added tests for scenarios that previously caused bugs. Let me know if there's anything else!

File: packages/next-query-params/src/NextQueryParamProvider.tsx [hunk 2]
```diff
@@ -11,7 +11,7 @@ type Props = Omit<
 
 function NextQueryParamProvider({children, ...rest}: Props) {
   const router = useRouter();
-  const match = router.asPath.match(/[^?]+/);
+  const match = router.asPath.match(/[^?|#]+/);
```
Comment (amannn) [inline packages/next-query-params/src/NextQueryParamProvider.tsx:14]: Is there a particular reason why you added the `|`? The match tries to capture the pathname, which excludes the query params and hash part. Shouldn't this work without the `|`?

**Review by amannn — COMMENTED** [review]
Great tests 👏

File: packages/next-query-params/src/NextQueryParamProvider.tsx (hunk 2 above)
Comment (amannn) [reply packages/next-query-params/src/NextQueryParamProvider.tsx:14]: Btw. possibly we could move the regex to a variable, so we don't have to duplicate it here and below.

Comment (chrisneven) [reply packages/next-query-params/src/NextQueryParamProvider.tsx]: The `|` is actually not necessary indeed, good one! I've also moved the regex to a variable :)

**Pull request merged**